### PR TITLE
Added download chart as image function

### PR DESCRIPTION
**Pull request by n3rsti**
I added download chart as image function in js. Function is async setTimeout with 100ms timeout, because otherwise it creates url of not existing chart before async ajax request. I added 3 anchors with buttons inside in graph.html.

I also edited row under "Edits over time" chart in style.css because without that, It would display buttons in row, not in column like the rest of chart buttons.

### CODE REVIEW

File: hashtagsv2/graphs/templates/graphs/graph.html [hunk 1]
```diff
@@ -153,18 +166,25 @@ <h1>
...
+                    {% comment %} Translators: This labels a button to download the data in JPG format. {% endcomment %}
+                    {% trans "Download JPG" %}
+                </button>
+            </a>
         </div>
     </div>
     <hr>
     <script> var url_param = '{{ query_string }}'; </script>
     <script src="{% static 'js/graph.js' %}"></script>
+    <script src="{% static 'js/tests.js' %}"></script>
```
Comment (Jain-Aditya) [inline hashtagsv2/graphs/templates/graphs/graph.html]: `tests.js` don't exist yet, so it will throw an error.

File: hashtagsv2/graphs/tests.py [hunk 2]
```diff
@@ -3,6 +3,7 @@
 
 from django.test import TestCase, RequestFactory
 
+
```
Comment (Jain-Aditya) [inline hashtagsv2/graphs/tests.py]: Looks like you have added an empty line here.

File: hashtagsv2/hashtags/static/js/graph.js [hunk 3]
```diff
@@ -118,4 +123,16 @@ $.ajax({
...
+    let projects_link = document.getElementById("projectsChart").toDataURL("image/jpeg");
+    let users_link = document.getElementById("usersChart").toDataURL("image/jpeg")
+    let edits_link = document.getElementById("timeChart").toDataURL("image/jpeg")
+    document.getElementById("projects_url").href = projects_link;
+    document.getElementById("projects_url").setAttribute('download', 'top_projects.jpg')
+    document.getElementById("users_url").href = users_link;
+    document.getElementById("users_url").setAttribute('download', 'top_users.jpg')
+    document.getElementById("time_url").href = edits_link;
+    document.getElementById("time_url").setAttribute('download', 'edits_over_time.jpg')
+}, 100);
```
Comment (Jain-Aditya) [inline hashtagsv2/hashtags/static/js/graph.js]: Each chart may take different time to load and also the charts are asynchronous. So, setting a time explicitly to call the function is not a good idea and may result in an inappropriate download. Instead you can include `animation` option in each chart such that each function is called after the animation completes.

For example, for the Users chart, you can add an option:

`animation: {
                        onComplete: function(){
                            let users_link = document.getElementById("usersChart").toDataURL("image/jpeg");
                            document.getElementById("users_url").href = users_link;
                            document.getElementById("users_url").setAttribute('download', 'top_users.jpg');
                        }
         }`

and simliarly you can do for all the three charts.

**Review by n3rsti — COMMENTED** [review]
I fixed problems included in pr comments

File: hashtagsv2/hashtags/static/js/graph.js [hunk 4]
```diff
@@ -34,15 +34,27 @@ else{
                             }
                         }]
                     },
+                    animation: { 
+                        onComplete: function(){ 
+                            let users_link = document.getElementById("projectsChart").toDataURL("image/jpeg");
```
Comment (Jain-Aditya) [inline hashtagsv2/hashtags/static/js/graph.js]: Just a small change regarding the variable naming. Could this be projects_link?

File: hashtagsv2/hashtags/static/js/graph.js [hunk 5]
```diff
@@ -111,7 +130,14 @@ $.ajax({
                         }
                     }]
                 },
-                maintainAspectRatio: false
+                maintainAspectRatio: false,
+                animation: { 
+                    onComplete: function(){ 
+                        let users_link = document.getElementById("timeChart").toDataURL("image/jpeg");
```
Comment (Jain-Aditya) [inline hashtagsv2/hashtags/static/js/graph.js]: Also here, Could it be time_link or anything else?

**Comment by Jain-Aditya** [comment]
Thanks! Everything works perfect as intended.

Regarding the tests, I think the best way to install Qunit is through node. Could you give it a try?

**Comment by n3rsti** [comment]
Yes, I will try.

**Comment by n3rsti** [comment]
I think I need help. I tried to test that. I couldn't get, how can I check if image from url exist, but I tried to compare chart.toDataUrl() with anchor href with
```javascript
QUnit.test( "URLs generated", function( assert ) {
        assert.equal( document.getElementById("users_url").href, document.getElementById("usersChart").toDataURL("image/jpeg") , "Urls succesfully generated" );
})
```
and I realised this test might not be sufficient. When I tried to install and run tests in docker script file, qunit command freezes whole docker-compose up --build -d process, so I did that in browser with jquery qunit package included in script src tag.


I found out that QUnit in browser tries to do all the tests at page start/refresh and because of that test written above couldn't work, because html document wasn't rendered yet, so I tried to do testing function

```javascript
function unit_test(){
    QUnit.test( "URLs generated", function( assert ) {
        assert.equal( document.getElementById("users_url").href, document.getElementById("usersChart").toDataURL("image/jpeg") , "Urls succesfully generated" );
    })
};
```

and put unit_test() function execute in graph.js when charts are generated (in complete function) and it worked, because test where succesful, but when QUnit doesn't find any tests at the beggining, It throws errors (Error: No tests were run.). I really want to solve that problem, but I can't find any idea how to test dynamically rendered content.

**Comment by n3rsti** [comment]
I just don't know how tests can be run with dynamically rendered content and I couldn't find any articles about testing dynamically rendered html document and I didn't find any mentions about that in QUnit docs.

**Comment by Jain-Aditya** [comment]
Now that the main task is completed, We can merge this PR. We can probably make a different task for testing the image download.
But before that please change the `users_link` variable name in the two functions as I have mentioned in the above comments. Also add a line of comment to each JS function you have written. After that, this task will be resolved.

**Comment by n3rsti** [comment]
It should be fine now.

**Comment by Jain-Aditya** [comment]
Looks good! Working as intended.

**Pull request merged**